MicroBenchmarks/LoopVectorization/ConditionalScalarAssignment.cpp

```diff
@@ -0,0 +1,222 @@
+  // Pick a default value that's out of range of the uniform distribution
+  // created for 'A', 'B', and 'C' in init_data below.
+  T ResultA = 101;
+  T ResultB = 101;
+  T ResultC = 101;
+#pragma clang loop vectorize(disable) interleave(disable)
+  for (unsigned i = 0; i < ITERATIONS; i++) {
+    if (A[i] > Threshold)
+      ResultA = A[i];
+    if (B[i] > Threshold)
+      ResultB = B[i];
+    if (C[i] > Threshold)
+      ResultC = C[i];
+  }
+
+  return ResultA ^ ResultB ^ ResultC;
+}
+
+// Find the last element in A above the given threshold,
+// with default loop vectorization settings.
+template <typename T>
+static T run_csa_with_arith_autovec(T *A, T *B, T *C, T Threshold) {
+  // Pick a default value that's out of range of the uniform distribution
+  // created for 'A' in init_data below.
+  T Result = 101;
+  for (unsigned i = 0; i < ITERATIONS; i++) {
+    // Do some work to make the difference noticeable
+    C[i] = A[i] * 13 + B[i] * 5;
+    if (A[i] > Threshold)
+      Result = A[i];
+  }
+
+  return Result;
+}
+
+// Find the last element in A above the given threshold,
+// with loop vectorization disabled.
+template <typename T>
+static T run_csa_with_arith_novec(T *A, T *B, T* C, T Threshold) {
+  // Pick a default value that's out of range of the uniform distribution
+  // created for 'A' in init_data below.
+  T Result = 101;
+#pragma clang loop vectorize(disable) interleave(disable)
+  for (unsigned i = 0; i < ITERATIONS; i++) {
+    // Do some work to make the difference noticeable
+    C[i] = A[i] * 13 + B[i] * 5;
+    if (A[i] > Threshold)
+      Result = A[i];
+  }
+
+  return Result;
+}
+
+// Initialize arrays A, B, and C with random numbers
+template <typename T> static void init_data(T *A, T* B, T *C) {
+  std::uniform_int_distribution<T> dist(0, 100);
+  std::mt19937 rng(12345);
+  for (unsigned i = 0; i < ITERATIONS; i++) {
+    A[i] = dist(rng);
+    B[i] = dist(rng);
+    C[i] = dist(rng);
+  }
+}
+
+// Benchmark auto-vectorized version.
+template <typename T>
+static void __attribute__((always_inline))
+benchmark_csa_autovec(benchmark::State &state, CSAFunc<T> VecFn,
+                      CSAFunc<T> NoVecFn, T Threshold) {
+  std::unique_ptr<T[]> A(new T[ITERATIONS]);
+  std::unique_ptr<T[]> B(new T[ITERATIONS]);
+  std::unique_ptr<T[]> C(new T[ITERATIONS]);
+  init_data(&A[0], &B[0], &C[0]);
+
+#ifdef BENCH_AND_VERIFY
+  // Verify the vectorized and un-vectorized versions produce the same results.
+  {
+    T VecRes = VecFn(&A[0], &B[0], &C[0], Threshold);
+    T NoVecRes = NoVecFn(&A[0], &B[0], &C[0], Threshold);
+    // We're only interested in whether the conditional assignment results
+    // were the same.
+    if (VecRes != NoVecRes) {
+      std::cerr << "ERROR: autovec result different to scalar result; "
+                << VecRes << " != " << NoVecRes << "\n";
+      exit(1);
+    }
+  }
+#endif
+
+  for (auto _ : state) {
+    VecFn(&A[0], &B[0], &C[0], Threshold);
```

Comment:
I am not sure this is working as expected. I think we need something like below to make sure the CAS result is used:

```
-    NoVecFn(&A[0], &B[0], &C[0], Threshold);
+    auto Res = NoVecFn(&A[0], &B[0], &C[0], Threshold);
+    benchmark::DoNotOptimize(Res);
```

Without a use of the result, compiler is probably able completely remove the variantst that don't have stores in the loop  and also remove  the unused CAS chain after inlining?